### PR TITLE
acme: Determine offline whether renewal is due

### PR DESCRIPTION
**Pull request by roberth**
###### Motivation for this change

Alleviate #85794 by using offline operations unless a renewal is required.
The acme service units will only fail if they actually need network access and are unlucky with the network during switching.

NOTE: currently the acme test fails (#114751) but if I exclude the previously offending cases, the test suite still succeeds.

I have also tested both the skipping and renewal behavior on a real system with the production Let's Encrypt service.

cc @NixOS/acme 

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/security/acme.nix [hunk 1]
```diff
@@ -274,6 +274,31 @@ let
...
+                  | sed -e 's/^notAfter=//'
+          )"
+          [[ -n "$expiration_line" ]]
+
+          expiration_date="$(date -d "$expiration_line" +%s)"
+          now="$(date +%s)"
+          expiration_s=$[expiration_date - now]
+          expiration_days=$[expiration_s / (3600 * 24)]   # rounds down
+
+          [[ 1 = $[expiration_days > ${toString cfg.validMinDays} ] ]]
```
Comment (m1cr0man) [inline nixos/modules/security/acme.nix:299]: Why not just `[[ $expiration_days -ge ${toString cfg.validMinDays} ]]`?

**Review by m1cr0man — APPROVED** [review]
Looks good. Small comment on the complexity of one of the conditional checks. I think using some if statements instead of relying on returncode + pipefail behaviour (which not everyone may understand) would be worth the extra complexity for better readability. I'll leave it up to you.

File: nixos/modules/security/acme.nix (hunk 1 above)
Comment (roberth) [reply nixos/modules/security/acme.nix]: My head was stuck between `$[ ]`s. I've applied it, thanks!

**Comment by roberth** [comment]
Initially I wrote the expiry function with explicit conditionals and custom error messages, but that did not improve readability of the log because of the duplication with `set -x` diagnostics which were turned on previously.

I've clarified the reliance on `set -e` with a little comment that I believe should be enough to point newcomers in the right direction while not adding too much noise.

**Comment by m1cr0man** [comment]
Can we get a merge on this @NixOS/acme ? I don't see any issues with it.

@roberth Were you able to run the test suite on it?

**Comment by arianvp** [comment]
Hmmm it's fine; but also bit of an ugly workaround what is in essence a bug in the NixOS activation logic.

**Review by arianvp — APPROVED** [review]

**Pull request merged**